### PR TITLE
feat(schema): basic framework of schema

### PR DESCRIPTION
**Pull request by noneback**

### CODE REVIEW

File: pkg/schema.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+func (bs *BSchema) FieldAt(pos int) Field {
+	return bs.fields[pos]
+}
+
+func (bs *BSchema) encodeKey() []byte {
+	return []byte(bs.namespace + separator + bs.name)
+}
+
+func (bs *BSchema) encodeVal() []byte {
+	header := []byte(bs.namespace + separator + bs.name + separator)
```
Comment (WenyXu) [inline pkg/schema.go:39]: double check it later

**Comment by WenyXu** [comment]
1. Put the updating dependencies into an independent pull request
2. Remove the irrelevant codes
4. Don't touch other irrelevant codes

File: pkg/schema.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+const (
+	SchemaTypePolicy = 0
+	SchemaTypeGroup  = 1
+)
+
+// BSchema represents a schema of a model, managing a table.
+type BSchema struct {
+	typ       SchemaType // policy or group
+	name      string
```
[the remark this answers is not in the document]
Comment (noneback) [reply pkg/schema.go:13]: You mean change `typ` to byte array or string type to byte array ?

File: pkg/codec/codec.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	Separator = []byte{byte(0)}
+)
+
+type encoder interface {
+	encodeKey() []byte
+	encodeVal() []byte
+}
+
+type decoder interface {
+	decode([]byte) interface{}
```
Comment (WenyXu) [inline pkg/codec/codec.go:15]: unexported function

File: pkg/codec/codec.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package codec
+
+const ()
+
+var (
+	Separator = []byte{byte(0)}
+)
+
+type encoder interface {
+	encodeKey() []byte
```
Comment (WenyXu) [inline pkg/codec/codec.go:10]: unexported function

File: pkg/codec/codec.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	encodeVal() []byte
```
Comment (WenyXu) [inline pkg/codec/codec.go:11]: unexported function

File: pkg/schema/schema.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (bs *BSchema) encodeKey() []byte {
+	buf := make([]byte, 0, len(bs.namespace)+1+len(bs.name))
+	buf = appendBytes(buf, bs.namespace, codec.Separator, bs.name)
+	return buf
+}
+
+func (bs *BSchema) encodeVal() []byte {
+	buf := make([]byte, 0) // TODO: optimize later
```
Comment (WenyXu) [inline pkg/schema/schema.go:46]: The exact size is known

File: pkg/schema/schema.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func appendBytes(buf []byte, elem ...[]byte) []byte {
+	for _, b := range elem {
+		buf = append(buf, b...)
+	}
+	return buf
+}
+
+func (bs *BSchema) decode([]byte) {}
```
Comment (WenyXu) [inline pkg/schema/schema.go:61]: unimplemented

File: pkg/schema/schema.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package schema
+
+import "github.com/casbin-mesh/neo/pkg/codec"
+
+type SchemaType byte
+
+const (
+	SchemaTypePolicy = 0
```
Comment (WenyXu) [inline pkg/schema/schema.go:8]: Can we use the null terminator here? It seems problematic.

File: pkg/schema/field.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package schema
+
+type FieldType byte
+
+const (
+	FieldTypeBinary = 1
```
Comment (WenyXu) [inline pkg/schema/field.go:6]: let's follow the bson type definition. https://github.com/casbin-mesh/neo/blob/feat/btuple-weny/pkg/primitive/bsontype/bsontype.go

**Review by WenyXu — CHANGES_REQUESTED** [review]
1. Implement the unimplemented functions
2. Add the tests.

**Review by WenyXu — APPROVED** [review]

**Pull request merged**